cores/arduino/stm32/LowPower.c

```diff
@@ -0,0 +1,289 @@
+    case 4 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN4
+#ifdef PWR_WAKEUP_PIN5
+    case 5 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN5
+    default :
+      break;
+  }
+}
+
+/**
+  * @brief  Enable the sleep mode.
+  * @param  None
+  * @retval None
+  */
+void LowPower_sleep(){
+  /*Suspend Tick increment to prevent wakeup by Systick interrupt.
+    Otherwise the Systick interrupt will wake up the device within 1ms (HAL time base)*/
+  HAL_SuspendTick();
+
+  /* Enter Sleep Mode , wake up is done once User push-button is pressed */
+  HAL_PWR_EnterSLEEPMode(PWR_MAINREGULATOR_ON, PWR_SLEEPENTRY_WFI);
+
+  /* Resume Tick interrupt if disabled prior to SLEEP mode entry */
+  HAL_ResumeTick();
+
+  if (WakeUpUartCb != NULL) {
+    WakeUpUartCb();
+  }
+}
+
+/**
+  * @brief  Enable the stop mode.
+  * @param  None
+  * @retval None
+  */
+void LowPower_stop(){
+#if STM32L4xx
+  if (WakeUpUart != NULL) {
+    HAL_UARTEx_EnableStopMode(WakeUpUart);
+  }
+#endif
+
+#ifdef STM32L0xx
+  /* Enable Ultra low power mode */
+  HAL_PWREx_EnableUltraLowPower();
+
+  /* Enable the fast wake up from Ultra low power mode */
+  HAL_PWREx_EnableFastWakeUp();
+
+  /* Select HSI as system clock source after Wake Up from Stop mode */
+  __HAL_RCC_WAKEUPSTOP_CLK_CONFIG(RCC_STOP_WAKEUPCLOCK_HSI);
+#endif
+
+  // Enter Stop mode
+  HAL_PWR_EnterSTOPMode(PWR_LOWPOWERREGULATOR_ON, PWR_STOPENTRY_WFI);
+
+  // Exit Stop mode reset clocks
+  SystemClock_Config();
+
+#if STM32L4xx
+  if (WakeUpUart != NULL) {
+    HAL_UARTEx_DisableStopMode(WakeUpUart);
+
+    if (WakeUpUartCb != NULL) {
+      WakeUpUartCb();
+    }
+  }
+#endif
+}
+
+/**
+  * @brief  Enable the standby mode. The board reset when leaves this mode.
+  * @param  None
+  * @retval None
+  */
+void LowPower_standby(){
+#ifdef STM32L0xx
+  /* Enable Ultra low power mode */
+  HAL_PWREx_EnableUltraLowPower();
+
+  /* Enable the fast wake up from Ultra low power mode */
+  HAL_PWREx_EnableFastWakeUp();
+#endif
+
+  HAL_PWR_EnterSTANDBYMode();
+}
+
+/**
+  * @brief  Enable the shutdown mode.The board reset when leaves this mode.
+  * @param  None
+  * @retval None
+  */
+void LowPower_shutdown(){
+#ifdef STM32L4xx
+  HAL_PWREx_EnterSHUTDOWNMode();
+#endif
+}
+
+/**
+  * @brief  Configure the UART as a wakeup source. A callback can be called when
+  *         the chip leaves the low power mode. See board datasheet to check
+  *         with which low power mode the UART is compatible.
```

Comment:
from the above implementation we can state that this is not supported in standby nor shutdown modes